### PR TITLE
add schema parameter to table provider factory create method

### LINKED ISSUE #4142: `TableProviderFactory::create` should have `Optional<DFSchemaRef>` parameter
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`datafusion::datasource::datasource::TableProviderFactory::create` should have `Optional<DFSchemaRef>` a parameter to support statements with table DDL e.g: `CREATE EXTERNAL TABLE demo_table (c1 integer, c2 integer) STORED AS ...` ?

**Describe the solution you'd like**

add `schema: Optional<DFSchemaRef>` to method signature

**Describe alternatives you've considered**

use `schema: DFSchemaRef` as parameter, which can be empty. I've believe version with Optional make more sense, as it is easier to understand. 

**Additional context**

### PR DESCRIPTION
**Pull request by milenkovicm**
# Which issue does this PR close?

Closes #4142.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

`TableProviderFactory::create` is a public interface and change will be backward incompatible, but as this interface is in state of flux recently it would make sense to break it now

### CODE REVIEW

**Comment by milenkovicm** [comment]
thanks @andygrove, 
I've added test which should check if table schema is according to ddl provided in the crate statement. Let me know if you have some other cases to test.

**Review by andygrove — APPROVED** [review]

**Comment by andygrove** [comment]
@avantgardnerio may be interested in reviewing this as well

**Review by avantgardnerio — APPROVED** [review]

File: datafusion/core/tests/sql/create_drop.rs [hunk 1]
```diff
@@ -387,6 +387,36 @@ async fn create_custom_table() -> Result<()> {
...
+    let schema = cat.schema("public").unwrap();
+
+    let exists = schema.table_exist("dt");
+    assert!(exists, "Table should have been created!");
+
+    let table_schema = schema.table("dt").unwrap().schema();
+
+    assert_eq!(3, table_schema.fields().len());
+
+    assert_eq!(&DataType::Int32, table_schema.field(0).data_type());
```
Comment (alamb) [inline datafusion/core/tests/sql/create_drop.rs:413]: 👍

**Review by alamb — COMMENTED** [review]
Thanks -- looks great to me @milenkovicm 

Looks like there are now some conflicts but otherwise this PR is ready to go

**Comment by milenkovicm** [comment]
thanks @alamb, looks like #4126 did most of the job I was intending with this pr, schema can be extracted from `cmd` directly. so this pr at the end contributes test for part of #4126

File: datafusion/core/tests/sql/create_drop.rs [hunk 2]
```diff
@@ -387,6 +387,36 @@ async fn create_custom_table() -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn create_external_table_with_ddl() -> Result<()> {
+    let mut table_factories: HashMap<String, Arc<dyn TableProviderFactory>> =
+        HashMap::new();
+    table_factories.insert("mocktable".to_string(), Arc::new(TestTableFactory {}));
```
Comment (alamb) [inline datafusion/core/tests/sql/create_drop.rs:394]: this is so cool

**Review by alamb — APPROVED** [review]
Thanks @milenkovicm

**Pull request merged**